### PR TITLE
WIP, NEP: add dtype design NEP - DtypeMeta, Dtype, and hierarchy

### PR DESCRIPTION
**Pull request by mattip**
Replaces PR #12630 after comments there.

### CODE REVIEW

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+manipulated, printed, and used in loops. In this NEP, **dtype** is used as the
+name of the class, and **descriptor** is used as the name of the instance. A
+descriptor is what is held by a ndarray object, in addition to a block of
+memory and information about the strides and shape of elements in that memory.
+Creating an instance of ``dtype`` *i.e.* ``a = np.dtype('int8')`` will result
+in a descriptor, which is a python object of type ``dtype``.
+
+The ``dtype`` obect instance ``a`` has attributes, among them ``a.type``, which
+is a class object. Instantiating that class object ``a.type(3)`` produces a
+Num`py `scalar <http://www.numpy.org/devdocs/reference/arrays.scalars.html>`_.
```
Comment (mattip) [inline doc/neps/nep-0029-dtype-as-type.rst]: wrong backtick

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+This NEP proposes a class heirarchy for dtypes. The ``np.dtype`` class will
```
Comment (mattip) [inline doc/neps/nep-0029-dtype-as-type.rst]: hierarchy spelled wrong (here and below), also abstrace->abstract

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+``cast``). 
+
+Downstream users of NumPy can subclass these type classes. Creating a categorical
+dtype would look like this (without error checking for out-of-bounds values)::
+
+    class Colors(Dtype):
+        itemsize = 8
+        colors = ['red', 'green', 'blue']
+        def get_format_function(self, data, **options):
+            class Format():
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: The purpose of a categorical dtype, as far as I can tell, is to give textual meaning to the values in the ndarray. The only place that has meaning is when printing them.

File: doc/neps/nep-0029-dtype-as-type.rst (hunk 1 above)
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: fixed

File: doc/neps/nep-0029-dtype-as-type.rst (hunk 2 above)
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: fixed

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+.. _arrayprint: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/arrayprint.py#L418
+.. _PyArray_AdaptFlexibleDtype: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/src/multiarray/convert_datatype.c#L164 
+.. _PyArray_CanCastSafely: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/src/multiarray/convert_datatype.c#L398
+
+Overall Design
+--------------
+
+The ``Dtype`` class (and any subclass without ``itemsize``) is by definition an
+abstract base class. A metaclass ``DtypeMeta`` is used to add slots for
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: adopted

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+    class GenericDescr(Dtype, metaclass=DtypeMeta):
+        pass
+
+    class IntDescr(GenericDescr):
+        def __repr__(self):
+            # subclass of IntDescr
+            return f"dtype('{_kind_to_stem[self.kind]}{self.itemsize:d}')"
+
+        def get_format_function(self, data, **options):
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: formatting exists on dtypes, it is based on a `switch` statement to get the proper formatting function in arrayprint. I am proposing to make that a method-based lookup on the dtype.

File: doc/neps/nep-0029-dtype-as-type.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: Removed metaclass entirely

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                elif isinstance(obj, type) and issubclass(obj, np.generic):
+                    return dtype_scalar_dict[obj]
+                else:
+                    # Dtype('int8') or Dtype('S10') or record descr
+                    return create_new_descr(cls, *args, **kwargs)
+            else:
+                # At import, when creating Dtype and subclasses
+                return type.__call__(cls, *args, **kwargs)
+
+    class Dtype():
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: Removed metaclass entirely

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        def __new__(cls, *args, **kwargs):
+            # Do not allow creating instances of abstract base classes
+            if not hasattr(cls, 'itemsize'):
+                raise ValueError("cannot create instances of "
+                                 f"abstract class {cls!r}")
+            return super().__new__(cls, *args, **kwargs)
+
+    class GenericDescr(Dtype, metaclass=DtypeMeta):
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: ok

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    "dtype('uint8')"
+
+Each class will have its own set of ArrFuncs (``clip``, ``fill``,
+``cast``). 
+
+Downstream users of NumPy can subclass these type classes. Creating a categorical
+dtype would look like this (without error checking for out-of-bounds values)::
+
+    class Colors(Dtype):
+        itemsize = 8
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: Changed to "Plant" to avoid confusion

**Comment by mattip** [comment]
Simplified by removing the metaclass and extending `Dtype.__new__`. Fwiw, ``arrayformat``now works by method lookup on the dtype.

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+            obj = args[0]
+            if isinstance(obj, int):
+                return dtype_int_dict[obj]
+            elif isinstance(obj, type) and issubclass(obj, np.generic):
+                return dtype_scalar_dict[obj]
+            else:
+                # Dtype('int8') or Dtype('S10') or record descr
+                return create_new_descr(cls, *args, **kwargs)
+
+    class GenericDtype(Dtype):
```
Comment (mhvk) [inline doc/neps/nep-0029-dtype-as-type.rst]: There seems to be no benefit to having `GenericDtype` any more. I thought the argument before that you don't necessarily want to override what `int` means, but want to be sure that `itemsize` exists, was logical, so I would move some of the `__new__` above here (or stick with the metaclass, it really wasn't bad and could be used to auto-fill methods based on `itemsize`, etc.)

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+Additionally, dtype subclasses are passed through to the subclass ``__new__``
+
+A prototype without error checking, without options handling, and describing
+only ``np.dtype(np.uint8)`` together with an overridable ``get_format_function``
+for ``arrayprint`` looks like::
+
+    import numpy as np
+
+    class Dtype():
+        def __new__(cls, *args, **kwargs):
```
Comment (mhvk) [inline doc/neps/nep-0029-dtype-as-type.rst]: Thinking more, the metaclass does in fact seem the best idea, but it should check for the presence of `itemsize` on any *instances* not on the class, i.e., it should let `type` do its work, and then check that `itemsize` exists. This, of course, is just what `ABCmeta` does...

**Review by mhvk — COMMENTED** [review]
@mattip - a few more comments, though I must admit I'm confused. Mostly, I feel the examples are too abstract; is the goal actually to be able to subclass `np.dtype` with the NEP work in place? How would one actually put in the cast functions?

Maybe most relevant: I think it should be closer to an abstract base class, where the presence of `itemsize` only gets checked for the instance, not for the class.

File: doc/neps/nep-0029-dtype-as-type.rst [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+            # subclass of IntDescr
+            return f"dtype('{_kind_to_stem[self.kind]}{self.itemsize:d}')"
+
+        def get_format_function(self, data, **options):
+            # replaces switch on dtype found in _get_format_function
+            # (in arrayprint), **options details missing
+            from  numpy.core.arrayprint import IntegerFormat
+            return IntegerFormat(data)
+
+    class UInt8Dtype(IntDtype):
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply doc/neps/nep-0029-dtype-as-type.rst]: > though with the metaclass, as the check for itemsize needs to come after initialization

Dtypes will be immutable, so I'd expect this to go through `__new__`

[the remark this answers is not in the document]
Comment (eric-wieser) [reply doc/neps/nep-0029-dtype-as-type.rst]: Note with the `issubclass(np.dtype, type)` approach from the outdate NEP, this last line could look like:
```python
class uint8dtype(metaclass=UIntDtype, itemsize=8):
    pass
```

**Comment by mhvk** [comment]
In https://github.com/numpy/numpy/pull/12630#issuecomment-451336197, it was noted:
> Right now dtype objects already have this: `PyArray_ArrFuncs->setitem` converts a Python object to raw memory, and `PyArray_ArrFuncs->getitem` goes the other way. These are absolutely something we should have as (slot) methods on new-style dtypes

I think this goes to the crux of what I've been wondering about: how would this be done if I subclass `dtype`?

Note that I think it is absolutely fine for the actual work to proceed in steps, i.e., keep using `ArrFuncs` for now, but I think if subclassing is a goal, the NEP should address how those can make that work. Here, it would be particularly useful to have a more concrete example may be good: How would one really do your  `Plant`, i.e., do actual setting/getting (based on `enum`?). But perhaps it is better to do something a bit simpler and closer to binary, e.g., implementing something that uses `struct.pack/unpack` (better if it is for something that is not a dtype already, which leaves the pascal string format 'p'...). Or a `1-bit` format that uses `np.unpackbits`?

**Comment by njsmith** [comment]
In the long run we certainly want to support defining new dtypes in either C or Python, so it makes sense for a spec or prototype to discuss both.

But if the plan is to implement this in stages, then it makes sense for the first stage to only support subclassing in C, since that's where all the current dtypes are defined. Then stage 2 might be adding slot wrappers so that dtype methods are callable from python, and stage 3 would be adding the glue to actually define the methods in Python.

File: doc/neps/nep-0029-dtype-as-type.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattip) [reply doc/neps/nep-0029-dtype-as-type.rst]: Overriding any of the `PyArray_Descr` functions or ufuncs requires C code. I was looking for an easy way to demonstrate a subclass, but it seems I only confused reviewers.

**Comment by mattip** [comment]
If we are going with the approach that Dtype should be a metaclass, this PR should be closed in favor of #12630

**Comment by mattip** [comment]
Closing, it seems starting over will be easier than resuscitating this. Hopefully the discussion will have been useful as a starting point for a replacement.

**Pull request closed without merging**